agentlocal/mocks.go

```diff
@@ -20,6 +20,8 @@ import (
 	"github.com/percona/pmm/api/agentlocalpb"
 	"github.com/percona/pmm/api/agentpb"
 	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/percona/pmm-agent/common"
 )
 
 //go:generate mockery -name=supervisor -case=snake -inpkg -testonly
```

Comment:
I agree with this, @AlekSi what do you think?

Reply:
Mostly to make it more visible.
What would be the main file there: https://github.com/percona/pmm-managed/tree/PMM-2.0/services/agents ?

Reply:
but we don't use them only for mocks, do we?

Reply:
> // supervisor is a subset of methods of supervisor.Supervisor used by this package.
> // We use it instead of real type for testing and to avoid dependency cycle.

That really are only two reasons. In other words, it is not _something like_ Supervisor – it is either a real Supervisor or mock. We don't have other implementations of this interface. That's not an abstraction. If we could, we would use the real `supervisor.Supervisor` there.

Reply:
> What would be the main file there: https://github.com/percona/pmm-managed/tree/PMM-2.0/services/agents ?
Where it's used there it should be

Reply:
in this case in registry.go

Reply:
Ok, I'll try to explain how I'm thinking to get the more clear picture why I don't like this filename:

- I see the code. I'm looking at it, there is `mocks.go` file. 
- First I read filename and think:
 > "It called `mocks.go` so looks like this is only for testing and _it contains all mocks_ for this package."

- Next, I'm going inside this file and read the comment like above:
> // supervisor is a subset of methods of supervisor.Supervisor used by this package.
> // We use it instead of real type for testing **and to avoid dependency cycle**.

Hmmm... Wait a minute... This is not only for testing but also for `Dependency Injection`. 😕 
But wait, it's also doesn't contain mocks 😐, because mocks are placed in other files: `mock_supervisor_test.go`, `mock_client_test.go`.

It seems someone misled me. 😞This is the main reason I prefer to store interfaces in the place where it used.

PS. Also maybe these interfaces can be rethought later as described here: https://about.sourcegraph.com/go/idiomatic-go#interface-all-the-things

Reply:
@rnovikovP could you add this discussion to the next go-backend call?
@AlekSi @DexterHD let's merge it for now and make a new task if it needs after the discussion on the call, what do you think? I think it's out of the scope of this task.

Reply:
@BupycHuk No problem

Reply:
>It called mocks.go so looks like this is only for testing

In Go that would be `mocks_test.go`.

>This is the main reason I prefer to store interfaces in the place where it used.

The single language feature (interfaces) is used for several different things that better should not be mixed together. For example, "generics" and "sum types" can be implemented with interfaces. That's something like that – interface there is not an abstraction mechanism.

I don't think `registry.go` is a "main" file of "agents" package.
Do you have any ideas for renaming `mocks.go`? `deps.go` is… meh for me. Anything else?

Reply:
> I don't think registry.go is a "main" file of "agents" package.
We shouldn't set interfaces in the "main" file, we should set them where they are used.

Reply:
@AlekSi what do you think about skipping this discussion for now, because it's out of the scope of this task?

Reply:
Sure, I'm already looking into this PR. Sorry for not making that explicit.

Reply:
> In Go that would be mocks_test.go.

`*_test.go` is a suffix for files containing tests, not for all stuff related to testing. From my POV if it called `mocks` it for testing purposes, as mocks is testing terminology.

> I don't think registry.go is a "main" file of "agents" package.
> Do you have any ideas for renaming mocks.go? deps.go is… meh for me. Anything else?

In Go main name of the package is the same as the package name itself (and a package is [something like namespace](https://www.ardanlabs.com/blog/2013/07/how-packages-work-in-go-language.html)). E.g. for `package actions` it probably should be `actions.go`. You can find this pattern in a lot of codebases, also it used in the standart library. 

For example:

- https://github.com/golang/go/tree/master/src/io
- https://github.com/golang/go/blob/master/src/encoding/encoding.go
- https://github.com/golang/go/tree/master/src/crypto

Or in some well structured projects:
https://github.com/micro/go-micro/tree/master/codec
https://github.com/micro/go-micro/tree/master/broker

> That's something like that – interface there is not an abstraction mechanism.

It is an abstraction mechanism, as Go documentation says: 
> Interfaces in Go provide a way to specify the behavior of an object. 

This intersects in abstract data types: 
> In some object-oriented languages, especially those without full multiple inheritance, the term interface is used to define an abstract type that contains no data but defines behaviours as method signatures.

So there is an abstraction mechanism. But also there is [Generalization](https://golang.org/doc/effective_go.html#generality) mechanism.